### PR TITLE
Fix caching

### PR DESCRIPTION
**Pull request by adius**
We had a large dependency tree and it crashed everytime with stackoverflow error.
This was because of too many recursive calls of `getDeps`.
Also the whole method seemed to be pretty broken and did not return the correct dependencies.

### CODE REVIEW

File: lib/dynamic-cache.js [hunk 1]
```diff
@@ -3,13 +3,15 @@
...
+        values.push(obj[key]);
+      }
+    }
   }
-
-  return Object.keys(obj).map(function (key) {
-    return obj[key];
-  });
+  return values
 }
```
Comment (ForbesLindesay) [inline lib/dynamic-cache.js]: Why this change?  Your code is functionally identical to the shorter, simpler code that was there before?

File: lib/dynamic-cache.js [hunk 2]
```diff
@@ -35,15 +37,31 @@ DynamicCache.prototype.update = function (cb) {
       this._queue = [];
     }
   }.bind(this);
+
   var getDeps = function (filename) {
-    if (this._files[filename]) {
-      return [[filename]].concat(this._files[filename].map(getDeps)).reduce(function (a, b) {
```
Comment (ForbesLindesay) [inline lib/dynamic-cache.js]: The bug in this is actually just that it double counts things, leading to circular structures creating an infinite loop.

File: lib/dynamic-cache.js (hunk 1 above)
Comment (adius) [inline lib/dynamic-cache.js]: The performance was pretty bad so I tried to improve it. But you're right this probably doesn't change much.

File: lib/dynamic-cache.js (hunk 2 above)
Comment (adius) [inline lib/dynamic-cache.js]: Yeah that was definitely one problem.
But the whole map, reduce, concat made it also pretty slow.
My version is much faster in our setup

**Comment by ForbesLindesay** [comment]
I'm closing this as it has now been fixed by moving over to watchify.

Would you like to be added as a collaborator on this project, to help review and merge pull requests?

**Comment by adius** [comment]
If you think I can help, sure! I'd be happy to!
